File: app/src/main/java/com/example/tradeup/data/repository/RatingRepositoryImpl.java
package com.example.tradeup.data.repository;

import androidx.annotation.NonNull;
import com.example.tradeup.core.utils.Callback;
import com.example.tradeup.data.model.Rating;
import com.example.tradeup.data.source.remote.FirebaseRatingSource;

import java.util.List;
import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class RatingRepositoryImpl implements RatingRepository {

    private final FirebaseRatingSource firebaseRatingSource;

    @Inject
    public RatingRepositoryImpl(FirebaseRatingSource firebaseRatingSource) {
        this.firebaseRatingSource = firebaseRatingSource;
    }

    @Override
    public void submitRating(@NonNull Rating rating, final Callback<Void> callback) {
        firebaseRatingSource.submitRating(rating)
                .addOnSuccessListener(aVoid -> callback.onSuccess(null))
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getRatingsForUser(String userId, long limit, final Callback<List<Rating>> callback) {
        firebaseRatingSource.getRatingsForUser(userId, limit)
                .addOnSuccessListener(callback::onSuccess)
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getRatingForTransaction(String transactionId, String raterId, final Callback<Rating> callback) {
        firebaseRatingSource.getRatingForTransaction(transactionId, raterId)
                .addOnSuccessListener(callback::onSuccess) // Rating có thể là null
                .addOnFailureListener(callback::onFailure);
    }
}

File: app/src/main/java/com/example/tradeup/data/repository/RatingRepositoryImpl.kt
package com.example.tradeup.data.repository

import com.example.tradeup.data.model.Rating
import com.example.tradeup.data.source.remote.FirebaseRatingSource // Giả sử bạn sẽ tạo FirebaseRatingSource.kt
import javax.inject.Inject

class RatingRepositoryImpl @Inject constructor(
    private val firebaseRatingSource: FirebaseRatingSource
) : RatingRepository {

    override suspend fun submitRating(rating: Rating): Result<Unit> {
        // return firebaseRatingSource.submitRating(rating)
        TODO("Implement submitRating in FirebaseRatingSource and call it here")
    }

    override suspend fun getRatingsForUser(userId: String): Result<List<Rating>> {
        // return firebaseRatingSource.getRatingsForUser(userId)
        TODO("Implement getRatingsForUser in FirebaseRatingSource and call it here")
    }

    override suspend fun getRatingForTransaction(transactionId: String, raterId: String): Result<Rating?> {
        // return firebaseRatingSource.getRatingForTransaction(transactionId, raterId)
        TODO("Implement getRatingForTransaction in FirebaseRatingSource and call it here")
    }
    // ...
}